### PR TITLE
integration-tests: Add contract tests

### PR DESCRIPTION
**Pull request by mslipper**
This PR adds integration tests for basic smart contract interactions and swaps on Uniswap. Tested against both my local Hardhat node and the Kovan staging environment.

Note that in order to deploy the Uniswap contracts I had to change the Solidity compiler version in the integration tests to `^0.7.6`.

### CODE REVIEW

File: integration-tests/hardhat.config.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ const config: HardhatUserConfig = {
     timeout: isLiveNetwork() ? 300_000 : 75_000,
   },
   solidity: {
-    version: '0.8.9',
+    version: '0.7.6',
```
Comment (smartcontracts) [inline integration-tests/hardhat.config.ts:21]: I would recommend using hardhat's native system for applying multiple compiler types so we can use 0.8.9 (or whatever the latest is at any given time) for most contracts and only apply older versions to contracts that really must have it. You can find a guide here: https://hardhat.org/guides/compile-contracts.html#multiple-solidity-versions

Comment (smartcontracts) [reply integration-tests/hardhat.config.ts:21]: ^^ Will also cut down the diff in this PR.

Comment (mslipper) [reply integration-tests/hardhat.config.ts:21]: Will do. Was unaware of that feature, thanks for letting me know.

File: integration-tests/package.json [hunk 2]
```diff
@@ -31,6 +31,9 @@
     "@types/shelljs": "^0.8.8",
     "@typescript-eslint/eslint-plugin": "^4.26.0",
     "@typescript-eslint/parser": "^4.26.0",
+    "@uniswap/v3-core": "1.0.0",
+    "@uniswap/v3-periphery": "^1.0.1",
+    "@uniswap/v3-sdk": "^3.6.2-optimism-regenesis",
```
Comment (smartcontracts) [inline integration-tests/package.json:36]: Hmm is there a reason to use this specific version for the SDK?

File: integration-tests/test/contracts.spec.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+import {UniswapV3Deployer} from 'uniswap-v3-deploy-plugin/dist/deployer/UniswapV3Deployer';
+import {abi as FACTORY_ABI} from '@uniswap/v3-core/artifacts/contracts/UniswapV3Factory.sol/UniswapV3Factory.json';
+import {abi as POOL_ABI} from '@uniswap/v3-core/artifacts/contracts/UniswapV3Pool.sol/UniswapV3Pool.json';
+
+import {OptimismEnv} from './shared/env';
+import {FeeAmount, TICK_SPACINGS} from '@uniswap/v3-sdk';
+
+chai.use(solidity);
+
+export const getMinTick = (tickSpacing: number) => Math.ceil(-887272 / tickSpacing) * tickSpacing;
```
Comment (smartcontracts) [inline integration-tests/test/contracts.spec.ts:14]: Where do these functions come from?

File: integration-tests/test/contracts.spec.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    let pool: Contract;
+
+    before(async () => {
+      const tokenA = await Factory__ERC20.deploy(
+        100000000,
+        'OVM1',
+        8,
+        'OVM1'
+      );
+      await tokenA.deployed();
```
Comment (smartcontracts) [inline integration-tests/test/contracts.spec.ts:110]: First time I've seen this.. whoops. Better than doing `await tokenA.deployTransaction.wait()`.

File: integration-tests/test/contracts.spec.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+      const tokenB = await Factory__ERC20.deploy(
+        100000000,
+        'OVM2',
+        8,
+        'OVM2'
+      );
+      await tokenB.deployed();
+      tokens = [tokenA, tokenB];
+      tokens.sort((a, b) => {
```
Comment (smartcontracts) [inline integration-tests/test/contracts.spec.ts:119]: The fact that `.sort()` mutates the underlying array is irritating to me but at least this code is correct.

File: integration-tests/test/contracts.spec.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
+import {BigNumber, Contract, ContractFactory, utils, Wallet} from 'ethers';
```
Comment (tynes) [inline integration-tests/test/contracts.spec.ts:1]: I don't think this file is being linted

File: integration-tests/test/contracts.spec.ts (hunk 3 above)
Comment (tynes) [reply integration-tests/test/contracts.spec.ts:14]: Why exactly are these exported?

File: integration-tests/test/contracts.spec.ts (hunk 6 above)
Comment (smartcontracts) [reply integration-tests/test/contracts.spec.ts:1]: Good catch. We have an error in our linting step. I'm going to make a PR to fix this.

Comment (smartcontracts) [reply integration-tests/test/contracts.spec.ts:1]: https://github.com/ethereum-optimism/optimism/pull/1697

File: integration-tests/package.json (hunk 2 above)
Comment (mslipper) [reply integration-tests/package.json:36]: It's just the one that got automatically added. I'll remove the regenesis tag.

File: integration-tests/test/contracts.spec.ts (hunk 3 above)
Comment (mslipper) [reply integration-tests/test/contracts.spec.ts:14]: I thought they might be useful in other tests. They come from the Uniswap test suite. I'll unexport and add some comments around where I got them.

File: integration-tests/test/contracts.spec.ts (hunk 6 above)
Comment (mslipper) [reply integration-tests/test/contracts.spec.ts]: Fixed the lint issues manually.

**Review by smartcontracts — APPROVED** [review]
LGTM. Assuming this passes CI, feel free to merge.

**Pull request merged**